### PR TITLE
Fix spin-loop/cleanup failure mode within run loop

### LINKED ISSUE #23: Difference in error behavior when a job is undefined
Hi! We have migrated `DelayedJob` to `Delayed` and are running it.
While observing the worker, we came across an error and found a difference in behavior.

Here is the stack trace in the `last_error` column.

<details>

> Job failed to load: undefined class/module Sample. Handler: \"--- !ruby/object:Delayed::JobWrapper\
> job_data:\
>   job_class: Sample\
>   job_id: 0f9f613a-682b-439c-8169-d45a4984e8eb\
>   provider_job_id: \
>   queue_name: v2\
>   priority: \
>   arguments: &2 []\
>   executions: 0\
>   exception_executions: &1 {}\
>   locale: en\
>   timezone: UTC\
>   enqueued_at: '2023-01-20T11:51:44Z'\
> job: !ruby/object:Sample\
>   arguments: []\
>   job_id: 0f9f613a-682b-439c-8169-d45a4984e8eb\
>   queue_name: v2\
>   priority: \
>   executions: 0\
>   exception_executions: *1\
>   timezone: UTC\
>   provider_job_id: \
>   serialized_arguments: *2\
>   locale: en\
>   enqueued_at: '2023-01-20T11:51:44Z'\
> \"
> /usr/local/bundle/gems/delayed-0.4.0/lib/delayed/backend/base.rb:84:in `rescue in payload_object'
> /usr/local/bundle/gems/delayed-0.4.0/lib/delayed/backend/base.rb:81:in `payload_object'
> /usr/local/bundle/gems/delayed-0.4.0/lib/delayed/backend/base.rb:133:in `max_run_time'
> /usr/local/bundle/gems/delayed-0.4.0/lib/delayed/worker.rb:194:in `max_run_time'
> /usr/local/bundle/gems/delayed-0.4.0/lib/delayed/worker.rb:137:in `block in run'
> /usr/local/lib/ruby/3.0.0/benchmark.rb:308:in `realtime'
> /usr/local/bundle/gems/delayed-0.4.0/lib/delayed/worker.rb:136:in `run'
> /usr/local/bundle/gems/delayed-0.4.0/lib/delayed/worker.rb:210:in `block in run_job'
> /usr/local/bundle/gems/delayed-0.4.0/lib/delayed/lifecycle.rb:61:in `block in initialize'
> /usr/local/bundle/gems/delayed-0.4.0/lib/delayed/lifecycle.rb:66:in `execute'
> /usr/local/bundle/gems/delayed-0.4.0/lib/delayed/lifecycle.rb:40:in `run_callbacks'
> /usr/local/bundle/gems/delayed-0.4.0/lib/delayed/worker.rb:210:in `run_job'
> /usr/local/bundle/gems/delayed-0.4.0/lib/delayed/worker.rb:102:in `block (4 levels) in work_off'
> /usr/local/bundle/gems/delayed-0.4.0/lib/delayed/plugins/connection.rb:7:in `block (3 levels) in <class:Connection>'
> /usr/local/bundle/gems/activerecord-6.1.6.1/lib/active_record/connection_adapters/abstract/connection_pool.rb:462:in `with_connection'
> /usr/local/bundle/gems/delayed-0.4.0/lib/delayed/plugins/connection.rb:6:in `block (2 levels) in <class:Connection>'
> /usr/local/bundle/gems/delayed-0.4.0/lib/delayed/lifecycle.rb:79:in `block (2 levels) in add'
> /usr/local/bundle/gems/delayed-0.4.0/lib/delayed/lifecycle.rb:61:in `block in initialize'
> /usr/local/bundle/gems/delayed-0.4.0/lib/delayed/lifecycle.rb:79:in `block in add'
> /usr/local/bundle/gems/delayed-0.4.0/lib/delayed/lifecycle.rb:66:in `execute'
> /usr/local/bundle/gems/delayed-0.4.0/lib/delayed/lifecycle.rb:40:in `run_callbacks'
> /usr/local/bundle/gems/delayed-0.4.0/lib/delayed/worker.rb:121:in `run_thread_callbacks'
> /usr/local/bundle/gems/delayed-0.4.0/lib/delayed/worker.rb:101:in `block (3 levels) in work_off'
> /usr/local/bundle/gems/concurrent-ruby-1.1.10/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:352:in `run_task'
> /usr/local/bundle/gems/concurrent-ruby-1.1.10/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:343:in `block (3 levels) in create_worker'
> /usr/local/bundle/gems/concurrent-ruby-1.1.10/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:334:in `loop'
> /usr/local/bundle/gems/concurrent-ruby-1.1.10/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:334:in `block (2 levels) in create_worker'
> /usr/local/bundle/gems/concurrent-ruby-1.1.10/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:333:in `catch'
> /usr/local/bundle/gems/concurrent-ruby-1.1.10/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:333:in `block in create_worker'

</details>

This error occurs when a job is enqueued even though the class definition of the job does not exist in the worker.
(For example, this can occur when adding a new job definition and trying to synchronize the webapp, worker Deployment with ArgoCD.)

I noticed that some kind of error occurs when executing a job, whether it is a `DelayedJob` or `Delayed`, but the difference is that `Delayed` does not retry and the failed_at column is updated.

The situation of enqueuing a non-existent job is not good, but we expect the worker to retry and complete the job execution once the code is up-to-date.

I am assuming that this difference in behavior exists because of the difference in serialization and deserialization of jobs.

In the case of `Delayed`, the `job` key is added to the serialization string of the job, and an error seems to occur when deserializing it.
`DeserializationError` is handled so that it is not retried.
(In `DelayedJob`, there is no `job` key and no error occurs during deserialization.)

```
--- !ruby/object:Delayed::JobWrapper
job_data:
 job_class: Sample
 job_id: c5d50b73-c8c9-4df0-8ea3-d2f6ca45b85d
 provider_job_id: 
 queue_name: v2
 priority: 
 arguments: &2 []
 executions: 0
 exception_executions: &1 {}
 locale: en
 timezone: UTC
 enqueued_at: '2023-01-20T04:17:09Z'
job: !ruby/object:Sample
 arguments: []
 job_id: c5d50b73-c8c9-4df0-8ea3-d2f6ca45b85d
 queue_name: v2
 priority: 
 executions: 0
 exception_executions: *1
 timezone: UTC
 provider_job_id: 
 serialized_arguments: *2
 locale: en
 enqueued_at: '2023-01-20T04:17:09Z'
```

I assume that the job can be executed even if this job key does not exist.

Is there any intention behind this difference in specifications?

### PR DESCRIPTION
**Pull request by smudge**
This ensures that exceptions raised in thread callback hooks are rescued and properly mark jobs as failed.

This is also a good opportunity to change the `num` argument (of `work_off(num)`) to mean number of jobs (give or take a few due to `max_claims`), not number of iterations. Previously (before threading was introduced) I think it meant number of jobs (though jobs and iterations were 1:1). I would not have done this before the refactor, because there was no guarantee that one of `success` or `failure` would be incremented (the thread might crash for many reasons). Now, we only increment `success` and treat `total - success` as the "failure" number when we return from the method.

Fixes #23 and #41

This is also a prereq for a resolution I'm cooking up for #36

### CODE REVIEW

File: lib/delayed/worker.rb [hunk 1]
```diff
@@ -89,22 +89,17 @@ def on_exit!
...
-      num.times do
+      while total < num
         jobs = reserve_jobs
         break if jobs.empty?
 
+        total += jobs.length
         pool = Concurrent::FixedThreadPool.new(jobs.length)
         jobs.each do |job|
           pool.post do
-            run_thread_callbacks(job) do
```
Comment (smudge) [inline lib/delayed/worker.rb:101]: The explanation might make more sense if I put it here next to the related code.

Previously, if `run_thread_callbacks` crashed for any reason, neither `success` nor `failure` would be incremented. Furthermore, the job cleanup (down at the end of `run(job)`) would never occur, so attempts would not be incremented and run_at would not be bumped, making the job immediately available for pickup again.

This is what I mean by the "spinloop". A job gets picked up, its thread crashes, no cleanup occurs, it gets immediately picked up again, etc. Pushing the `run_thread_callbacks` call into `run_job` allows us to clean up the job if, e.g., it fails to deserialize, or if one of our callbacks fails to connect to a secondary resource (as was the case in #41).

File: lib/delayed/worker.rb [hunk 2]
```diff
@@ -114,38 +109,41 @@ def work_off(num = 100)
...
+        job_say job, format('COMPLETED after %.4f seconds', run_time)
       end
-      job_say job, format('COMPLETED after %.4f seconds', run_time)
       true # did work
     rescue DeserializationError => e
       job_say job, "FAILED permanently with #{e.class.name}: #{e.message}", 'error'
 
       job.error = e
       failed(job)
+      false # work failed
```
Comment (smudge) [inline lib/delayed/worker.rb:146]: This was a bug that improperly reported deserialization errors in the `success` number, which only impacted the logging output.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -8,7 +8,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        ruby: ['2.6', '2.7', '3.0', '3.1', '3.2']
+        ruby: ['2.7', '3.0', '3.1', '3.2']
```
Comment (smudge) [inline .github/workflows/ci.yml:11]: we can make this more formal, but I wanted to unblock this build for now, without the linter churn that comes from actually changing the min supported Ruby.

File: lib/delayed/worker.rb [hunk 4]
```diff
@@ -89,22 +89,17 @@ def on_exit!
     # Exit early if interrupted.
     def work_off(num = 100)
       success = Concurrent::AtomicFixnum.new(0)
-      failure = Concurrent::AtomicFixnum.new(0)
+      total = 0
 
-      num.times do
+      while total < num
```
Comment (samandmoore) [inline lib/delayed/worker.rb:94]: So I'm not sure how likely it is to matter but this change means that we will keep looping until we see num jobs which means that if the queue becomes empty before we hit num we will keep looping, right? 

Is that okay or desired?

**Review by samandmoore — COMMENTED** [review]
Change looks good as a solve for the issue. Just one question about the other behavior change!

File: lib/delayed/worker.rb (hunk 4 above)
Comment (samandmoore) [reply lib/delayed/worker.rb:94]: ah! i missed that. excellent.

**Review by samandmoore — APPROVED** [review]
domainlgtm

**Review by samandmoore — APPROVED** [review]
platformlgtm

**Pull request merged**